packages/go_router/lib/src/router.dart

```diff
@@ -82,26 +87,44 @@ class GoRouter extends ChangeNotifier with NavigatorObserver {
     }());
   }
 
-  /// The route information parser used by the go router.
-  late final GoRouteInformationParser routeInformationParser;
+  late final RouteConfiguration _routeConfiguration;
+  late final GoRouteInformationParser _routeInformationParser;
+  late final GoRouterDelegate _routerDelegate;
+  late final GoRouteInformationProvider _routeInformationProvider;
+
+  /// The router delegate. Provide this to the MaterialApp or CupertinoApp's
+  /// `.router()` constructor
+  // TODO(johnpryan): change type to RouterDelegate
+  GoRouterDelegate get routerDelegate => _routerDelegate;
+
+  /// The route information provider used by [GoRouter].
+  // TODO(johnpryan): change type to RouteInformationProvider
+  GoRouteInformationProvider get routeInformationProvider =>
+      _routeInformationProvider;
 
-  /// The router delegate used by the go router.
-  late final GoRouterDelegate routerDelegate;
+  /// The route information parser used by [GoRouter].
+  // TODO(johnpryan): change type to RouteInformationParser
```

Comment:
I feel we will still need to expose the GoRoute- delegates if customers want to create their own delegates to use with the GoRouter- delegates

Reply:
This class isn't currently part of the public API, but yes if we wanted to make it  configurable as the part of the constructor we would have to export it:


```dart
import 'package:go_router/go_router.dart'
class CustomMatchListParser extends MatchListParser {
  // ...
}

// ... 
  final _router = GoRouter(
    builder: CustomMatchListBuilder(),
    parser: CustomMatchListParser(),
    matcher: CustomMatchListMatcher(),
    redirector: CustomMatchListRedirector(),
    routes: [
      // ...
    ],
```

We could take it one step further and allow users to specify a custom implementation of RouteConfiguration.

Reply:
I mean unless we are never going to expose GoRouteInfromationParser/GoRouterDelegate, it doesn't make sense to change the type back to RouteInformationParser given that it is already exposed as GoRouteInformationParser. I think right now this package is in a weird state that some classes should have been exposed as public but don't, and it was too late to making it completely private because some other classes have exposed them as part of their APIs

Reply:
I think I misunderstood your question, sorry. There was a discussion about this in a previous review: https://github.com/flutter/packages/pull/2317#discussion_r921422617